File: Sources/LLMCore/ThinkTagFilter.swift
/// Strips `<think>…</think>` reasoning blocks from a streamed text feed.
///
/// Stateful so it survives chunk boundaries: a chunk may end mid-tag (`…<thi`)
/// or mid-block. Feed chunks in order via `feed(_:)`; the returned string is the
/// visible text to emit so far. Call `flush()` at end of stream to release any
/// text that was held back as a possible-but-incomplete tag.
///
/// Only the literal tags `<think>` / `</think>` are recognized (case-sensitive),
/// which is what current reasoning models emit. An unclosed `<think>` drops
/// everything to the end of the stream.
public struct ThinkTagFilter {
    private static let open = Array("<think>")
    private static let close = Array("</think>")

    private var insideThink = false
    /// Text held back because it might be the prefix of a tag split across chunks.
    private var pending: [Character] = []

    public init() {}

    /// Feed the next chunk; returns visible text ready to emit.
    public mutating func feed(_ chunk: String) -> String {
        pending.append(contentsOf: chunk)
        var out = ""

        while !pending.isEmpty {
            let tag = insideThink ? Self.close : Self.open
            if let range = firstMatch(of: tag, in: pending) {
                // Text before the tag: emit it (only when outside a think block).
                if !insideThink { out += String(pending[..<range.lowerBound]) }
                pending.removeSubrange(..<range.upperBound)
                insideThink.toggle()
            } else if let partialLen = longestSuffixPrefix(of: tag, in: pending) {
                // Tail of pending could be the start of `tag`; keep it, emit the rest.
                let emitEnd = pending.count - partialLen
                if !insideThink { out += String(pending[..<emitEnd]) }
                pending.removeSubrange(..<emitEnd)
                break
            } else {
                // No tag and no possible partial: emit everything (if visible).
                if !insideThink { out += String(pending) }
                pending.removeAll()
                break
            }
        }
        return out
    }

    /// End of stream: emit any held-back text that turned out not to be a tag.
    /// Anything still inside an unclosed think block is dropped.
    public mutating func flush() -> String {
        defer { pending.removeAll() }
        return insideThink ? "" : String(pending)
    }

    // MARK: - Matching helpers

    /// Index range of the first full occurrence of `needle` in `haystack`.
    private func firstMatch(of needle: [Character], in haystack: [Character]) -> Range<Int>? {
        guard needle.count <= haystack.count else { return nil }
        for start in 0...(haystack.count - needle.count) {
            if Array(haystack[start..<start + needle.count]) == needle {
                return start..<(start + needle.count)
            }
        }
        return nil
    }

    /// Longest k>0 where the last k chars of `haystack` equal the first k of `needle`.
    /// That suffix might grow into a full tag on the next chunk, so hold it back.
    private func longestSuffixPrefix(of needle: [Character], in haystack: [Character]) -> Int? {
        let maxK = min(needle.count - 1, haystack.count)
        var k = maxK
        while k > 0 {
            if Array(haystack.suffix(k)) == Array(needle.prefix(k)) { return k }
            k -= 1
        }
        return nil
    }
}

File: windows/AnywhereLLM.Core/ThinkTagFilter.cs
using System.Text;

namespace AnywhereLLM.Core;

/// Strips &lt;think&gt;…&lt;/think&gt; reasoning blocks from a streamed text feed.
///
/// Stateful so it survives chunk boundaries: a chunk may end mid-tag ("…&lt;thi") or
/// mid-block. Feed chunks in order via Feed(); the returned string is the visible
/// text to emit so far. Call Flush() at end of stream to release any text held back
/// as a possible-but-incomplete tag.
///
/// Only the literal tags &lt;think&gt; / &lt;/think&gt; are recognized (case-sensitive), which
/// is what current reasoning models emit. An unclosed &lt;think&gt; drops everything to
/// the end of the stream. Port of the Swift ThinkTagFilter (a class here, not a
/// struct, to avoid value-copy foot-guns over the mutable pending buffer).
public sealed class ThinkTagFilter
{
    private static readonly char[] Open = "<think>".ToCharArray();
    private static readonly char[] Close = "</think>".ToCharArray();

    private bool _insideThink;
    private readonly List<char> _pending = new();

    /// Feed the next chunk; returns visible text ready to emit.
    public string Feed(string chunk)
    {
        _pending.AddRange(chunk);
        var sb = new StringBuilder();

        while (_pending.Count > 0)
        {
            var tag = _insideThink ? Close : Open;
            int idx = FirstMatch(tag);
            if (idx >= 0)
            {
                if (!_insideThink) AppendRange(sb, 0, idx);
                _pending.RemoveRange(0, idx + tag.Length);
                _insideThink = !_insideThink;
            }
            else
            {
                int partial = LongestSuffixPrefix(tag);
                if (partial > 0)
                {
                    int emitEnd = _pending.Count - partial;
                    if (!_insideThink) AppendRange(sb, 0, emitEnd);
                    _pending.RemoveRange(0, emitEnd);
                    break;
                }
                if (!_insideThink) AppendRange(sb, 0, _pending.Count);
                _pending.Clear();
                break;
            }
        }
        return sb.ToString();
    }

    /// End of stream: emit any held-back text that turned out not to be a tag.
    /// Anything still inside an unclosed think block is dropped.
    public string Flush()
    {
        var s = _insideThink ? "" : new string(_pending.ToArray());
        _pending.Clear();
        return s;
    }

    private void AppendRange(StringBuilder sb, int start, int endExclusive)
    {
        for (int i = start; i < endExclusive; i++) sb.Append(_pending[i]);
    }

    /// Index of the first full occurrence of needle in _pending, or -1.
    private int FirstMatch(char[] needle)
    {
        if (needle.Length > _pending.Count) return -1;
        for (int start = 0; start <= _pending.Count - needle.Length; start++)
        {
            bool ok = true;
            for (int k = 0; k < needle.Length; k++)
                if (_pending[start + k] != needle[k]) { ok = false; break; }
            if (ok) return start;
        }
        return -1;
    }

    /// Longest k&gt;0 where the last k chars of _pending equal the first k of needle;
    /// that suffix might grow into a full tag next chunk, so hold it back. 0 = none.
    private int LongestSuffixPrefix(char[] needle)
    {
        int maxK = Math.Min(needle.Length - 1, _pending.Count);
        for (int k = maxK; k > 0; k--)
        {
            bool ok = true;
            for (int i = 0; i < k; i++)
                if (_pending[_pending.Count - k + i] != needle[i]) { ok = false; break; }
            if (ok) return k;
        }
        return 0;
    }
}
